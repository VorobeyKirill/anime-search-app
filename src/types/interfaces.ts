// hungarian notation tends ot be problematic, it doesn't really add much
export interface ISearchState {
  loading: boolean;
  error: null | { message: string };
}

export interface IAnimeItem {
  synopsis: string;
  mal_id: number;
  title: string;
  images: {
    jpg: {
      image_url: string;
      large_image_url: string;
    };
  };
}
