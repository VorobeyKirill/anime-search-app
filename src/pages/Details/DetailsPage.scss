.details__loading-indicator {
  display: block;
  text-align: center;
  margin-top: 30px;
  font-size: 2rem;
  font-weight: 700;
}

.details__error {
  margin: 30px auto;
  width: 60%;
  text-align: center;
  font-size: 1.25rem;
}

/* Desktop styles */
@media screen and (min-width: 768px) {
  .details-container {
    display: grid;
    grid-template-areas:
      'image divider title'
      'image divider description';
    grid-template-columns: 30% 1px 70%;
    grid-template-rows: auto 1fr;
    max-width: 800px;
    margin: 20px auto;
    gap: 16px;
  }

  .details__title {
    grid-area: title;
    font-size: 1.5rem;
  }

  .details__image {
    grid-area: image;
    width: 100%;
    border-radius: 10px;
  }

  .details__description {
    grid-area: description;
    width: 80%;
  }

  .details__divider {
    grid-area: divider;
    height: 100%;
    width: 1px;
    background-color: #ccc;
  }
}

/* Mobile styles */
@media (max-width: 768px) {
  .details-container {
    margin: 30px;
  }

  .details__title {
    text-align: center;
    font-size: 2rem;
  }

  .details__image {
    display: block;
    width: 60%;
    min-width: 280px;
    margin: 0 auto;
    border-radius: 10px;
  }

  .details__divider {
    display: none;
  }
}
