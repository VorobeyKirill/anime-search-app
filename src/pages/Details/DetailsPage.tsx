import { useQuery } from 'react-query';
import { useParams } from 'react-router-dom';
import { IAnimeItem } from '../../types/interfaces';

const ANIME_API = 'https://api.jikan.moe/v4/anime';

async function fetchAnimeDataById(
  animeId: string | undefined
): Promise<IAnimeItem> {
  const response = await fetch(`${ANIME_API}/${animeId}`);
  const animeData = await response.json();

  return animeData.data;
}

export const DetailsPage = () => {
  const { id } = useParams();
  const { data, isLoading, error } = useQuery(
    ['animeData'],
    () => fetchAnimeDataById(id),
    { enabled: !!id }
  );

  if (isLoading) {
    return <span>Loading...</span>;
  }

  if (error) {
    return (
      <div>
        <span>
          Oops! Something went wrong. Please, try to run search again...
        </span>
        <p>Error: {(error as { message: string }).message}</p>
      </div>
    );
  }

  return (
    <>
      {!!data && (
        <>
          <h2 className="details__title">{data.title}</h2>
          <img
            src={data?.images.jpg.image_url}
            alt={data.title + ' cover image'}
            className="details__image"
          />
          <p className="details__description">{data.synopsis}</p>
        </>
      )}
    </>
  );
};
