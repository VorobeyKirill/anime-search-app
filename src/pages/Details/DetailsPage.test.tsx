import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { DetailsPage } from './DetailsPage';
import { IAnimeItem } from '../../types/interfaces';

const mockedAnimeItem: IAnimeItem = {
  synopsis: 'synopsis',
  title: 'title',
  mal_id: 1,
  images: {
    jpg: {
      image_url: 'url',
      large_image_url: 'large_url',
    },
  },
};

jest.mock('react-query', () => ({
  ...jest.requireActual('react-query'),
  useQuery: () => ({
    data: mockedAnimeItem,
    isLoading: false,
    error: null,
  }),
}));

describe('DetailsPage', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    expect(render(<DetailsPage />)).toMatchSnapshot();
  });

  it('should not render loading indicator by default', async () => {
    const { container } = render(<DetailsPage />);

    expect(
      container.querySelector('.details__loading-indicator')
    ).not.toBeInTheDocument();
  });

  it('should not render error message by default', async () => {
    const { container } = render(<DetailsPage />);

    expect(container.querySelector('.details__error')).not.toBeInTheDocument();
  });

  it('should render anime title if animeData was returned from useQuery', async () => {
    const { container } = render(<DetailsPage />);

    expect(container.querySelector('.details__title')).toBeInTheDocument();
  });

  it('should render anime title with correct textContent if animeData was returned from useQuery', async () => {
    const { container } = render(<DetailsPage />);

    expect(container.querySelector('.details__title')?.textContent).toEqual(
      mockedAnimeItem.title
    );
  });

  it('should render anime image if animeData was returned from useQuery', async () => {
    const { container } = render(<DetailsPage />);

    expect(container.querySelector('.details__image')).toBeInTheDocument();
  });

  it('should render anime description if animeData was returned from useQuery', async () => {
    const { container } = render(<DetailsPage />);

    expect(
      container.querySelector('.details__description')
    ).toBeInTheDocument();
  });

  it('should render anime description with correct textContent if animeData was returned from useQuery', async () => {
    const { container } = render(<DetailsPage />);

    expect(
      container.querySelector('.details__description')?.textContent
    ).toEqual(mockedAnimeItem.synopsis);
  });
});
