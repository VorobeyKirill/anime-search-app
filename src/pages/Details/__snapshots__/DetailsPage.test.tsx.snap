// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPage should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="details-container"
      >
        <h2
          class="details__title"
        >
          title
        </h2>
        <img
          alt="title cover image"
          class="details__image"
          src="large_url"
        />
        <div
          class="details__divider"
        />
        <p
          class="details__description"
        >
          synopsis
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="details-container"
    >
      <h2
        class="details__title"
      >
        title
      </h2>
      <img
        alt="title cover image"
        class="details__image"
        src="large_url"
      />
      <div
        class="details__divider"
      />
      <p
        class="details__description"
      >
        synopsis
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
  // there's really no point in storing all these values in snapshots, it just makes them harder to reason about
}
`;
